commands/command_filter.go

```diff
@@ -0,0 +1,215 @@
+		// containing the bytes that we should write back out to Git.
+		_, err = to.Write(errors.GetContext(err, "bytes").([]byte))
+		return err
+	}
+
+	if err != nil {
+		Panic(err, "Error cleaning asset.")
+	}
+
+	tmpfile := cleaned.Filename
+	mediafile, err := lfs.LocalMediaPath(cleaned.Oid)
+	if err != nil {
+		Panic(err, "Unable to get local media path.")
+	}
+
+	if stat, _ := os.Stat(mediafile); stat != nil {
+		if stat.Size() != cleaned.Size && len(cleaned.Pointer.Extensions) == 0 {
+			Exit("Files don't match:\n%s\n%s", mediafile, tmpfile)
+		}
+		Debug("%s exists", mediafile)
+	} else {
+		if err := os.Rename(tmpfile, mediafile); err != nil {
+			Panic(err, "Unable to move %s to %s\n", tmpfile, mediafile)
+		}
+
+		Debug("Writing %s", mediafile)
+	}
+
+	_, err = cleaned.Pointer.Encode(to)
+	return err
+}
+
+func smudge(to io.Writer, reader io.Reader, filename string) error {
+	var pbuf bytes.Buffer
+	reader = io.TeeReader(reader, &pbuf)
+
+	ptr, err := lfs.DecodePointer(reader)
+	if err != nil {
+		// If we tried to decode a pointer out of the data given to us,
+		// and the file was _empty_, write out an empty file in
+		// response. This occurs because when the clean filter
+		// encounters an empty file, and writes out an empty file,
+		// instead of a pointer.
+		//
+		// TODO(taylor): figure out if there is more data on the reader,
+		// and buffer that as well.
+		if len(pbuf.Bytes()) == 0 {
+			if _, cerr := io.Copy(to, &pbuf); cerr != nil {
+				Panic(cerr, "Error writing data to stdout:")
+			}
+			return nil
+		}
+
+		return err
+	}
+
+	lfs.LinkOrCopyFromReference(ptr.Oid, ptr.Size)
+
+	cb, file, err := lfs.CopyCallbackFile("smudge", filename, 1, 1)
+	if err != nil {
+		Error(err.Error())
+	}
+
+	cfg := config.Config
+	download := lfs.FilenamePassesIncludeExcludeFilter(filename, cfg.FetchIncludePaths(), cfg.FetchExcludePaths())
+
+	if filterSmudgeSkip || cfg.Os.Bool("GIT_LFS_SKIP_SMUDGE", false) {
+		download = false
+	}
+
+	err = ptr.Smudge(to, filename, download, TransferManifest(), cb)
+	if file != nil {
+		file.Close()
+	}
+
+	if err != nil {
+		// Download declined error is ok to skip if we weren't requesting download
+		if !(errors.IsDownloadDeclinedError(err) && !download) {
+			LoggedError(err, "Error downloading object: %s (%s)", filename, ptr.Oid)
+			if !cfg.SkipDownloadErrors() {
+				// TODO: What to do best here?
+				os.Exit(2)
+			}
+		}
+
+		_, err = ptr.Encode(to)
+		return err
+	}
+
+	return nil
+}
+
+func filterCommand(cmd *cobra.Command, args []string) {
+	requireStdin("This command should be run by the Git filter process")
+	lfs.InstallHooks(false)
+
+	s := git.NewObjectScanner(os.Stdin, os.Stdout)
+
+	if err := s.Init(); err != nil {
+		ExitWithError(err)
+	}
+	if err := s.NegotiateCapabilities(); err != nil {
+		ExitWithError(err)
+	}
+
+Scan:
+	for s.Scan() {
+		var err error
+		var w io.Writer
+
+		req := s.Request()
+		if req == nil {
+			break
```

Comment:
This would only happen in case of an error, right?  If this is true, then maybe we should add a comment here? This error would kill the `git-lfs` process. This is good because `git` would detect that and restart `git-lfs`.